### PR TITLE
minimum mps support

### PR DESCRIPTION
**Pull request by robbiemu**
the quick-and-dirty version  

This lets VibeVoice run on Apple Silicon (without CUDA). It faithfully does the bare minimum: detects the Mac’s MPS backend, picks a safe dtype/attention combo, and starts the demo. Everything else (CLI flags, bfloat16, tests, diagnostics) was cut so the diff stays small and readable.

### How it works 
- Essentially the structure is the same, but less robust. 
- A helper (`vibevoice/utils/devices.py`) chooses the best device/attention stack.  
- `gradio_demo.py` and `inference_from_file.py` now call that helper; no `--device` or `--dtype` flags remain.  
- `flash-attn` is now an optional `[cuda]` extra, so pip install succeeds on macOS.  
- If Flash-Attention isn’t found, the code quietly falls back to SDPA.  

### What you lose  
- No runtime switches: power users can’t override the auto-detected settings.  
- MPS is locked to float16; faster bfloat16 support is gone.  
- Device-loading code is duplicated in both demo scripts.  
- No tests, no benchmark script, no diagnostic tool, no automatic env-vars.  
- The whole-repo linting fixes that accidentally rode along in the bigger PR (and maybe killed it) are also gone.

### Quick test  
Apple Silicon  
```
python demo/inference_from_file.py --model_path …
# should print “Device: mps” and emit audio
```

CUDA w/ flash-attn  
```
pip install -e .[cuda]
python demo/inference_from_file.py …
# should print “Attention: flash_attention_2”
```

CUDA w/o flash-attn  
```
pip uninstall flash-attn
python demo/inference_from_file.py …
# should fall back to SDPA
```

My 2 cents: This branch gives VibeVoice mps support without the end-user control features, etc, and leaves the chore of keeping it maintainable til another day. The full-featured PR (#38) is still around if you want the polished, future-proof solution.

### CODE REVIEW

**Comment by YaoyaoChang** [comment]
Key points look like:
* `flash_attention_2` -> `sdpa` : already auto-fallback in `inference_from_file.py `and `gradio_demo.py` (see https://github.com/microsoft/VibeVoice/commit/ac0104c65e816eadfc942764d17a512404bd5cc6 and https://github.com/microsoft/VibeVoice/commit/174e53fc04ba5eec33350f3958604555a5a64f88)

* `bfloat16` -> `float16` : is this required for MPS or just better?

* `inputs = {k: v.to(device) if torch.is_tensor(v) else v for k, v in inputs.items()}` :do we really need this for MPS? what does this line used for?

Any other important code for MPS? If not, please try to keep these functions concise, ideally within roughly 30 lines. If there are additional essentials, could you clarify them and still aim for the minimum necessary changes?

**Comment by YaoyaoChang** [comment]
PR [#67](https://github.com/microsoft/VibeVoice/pull/67) added MPS support. 
Many thanks again for your help! Any further advice would be greatly appreciated.”

**Comment by robbiemu** [comment]
> PR [#67](https://github.com/microsoft/VibeVoice/pull/67) added MPS support. Many thanks again for your help! Any further advice would be greatly appreciated.”

I'm glad you all have added support. 🚀 If you want anything else from my fork like a torch compile PR I would be happy to give you that too (but it is straightforward and you might prefer to do it yourself for maximum control).

Looking at the line counts in Saturday's pr and the one you referenced, it turns out to be just about identical to this one in terms of complexity. But #67 used your Saturday PR's device code. It makes the trade off for device and model loading in the other direction, instead of maintaining a modular system it bundles in the model logic (for example in gradio, this matters here because it simplifies device detection at the cost of duplicating code).

> flash_attention_2 -> sdpa

Yes I guess that is correct. it was necessary when I initially made the branch taht led to this, but I didn't really factor in the changes to main that occurred afterwards, my bad.

> inputs = {k: v.to(device) if torch.is_tensor(v) else v for k, v in inputs.items()} :do we really need this for MPS? what does this line used for?

absolutely you do. This is done now in the main branch here, using a for loop. It loops through the inputs dictionary and moves every PyTorch tensor to the correct device. The only difference is that main's version uses a for loop, whereas this PR used a more compact dictionary comprehension to achieve the same result.

VibeVoiceProcessor, when you call it with return_tensors="pt", the inputs dictionary it returns contains the following keys:

  Tensor Values:
   * input_ids: token IDs representing the formatted text prompt.
   * attention_mask
   * speech_input_mask: boolean tensor that identifies which of the input_ids are special
     placeholders for speech
   * speech_tensors: tensor containing a padded numerical representation of the audio voice
     samples
   * speech_masks: boolean mask the padded portions of speech_tensors

  Non-Tensor Values:
   * parsed_scripts: A standard Python list of tuples, where each tuple contains a speaker ID
     and their line of dialogue.
   * all_speakers_list: A list of lists, containing the unique speaker IDs found in each
     script.

 The presence of these non-tensor lists is the why the `if torch.is_tensor(v)` check is required.

> bfloat16 -> float16 : is this required for MPS or just better?

I like the fp32 default, like I was saying, fp16 is considered stable for mps (as I understand it) in torch but it still leads to unusable inference at about the same number of tokens here as bf16, which on my machine is ~600 tokens (I wasn't seeing NaNs, unlike the same issue in diaTTS, but the behavior was that rendered audio cut short) -- so fp32 default I guess makes sense. I defaulted to fp16 for speed, assuming a user needing stability could just change the dtype in my own.  

bf16 for mps is considered a work in progress in torch, again as I understand it.

**Comment by YaoyaoChang** [comment]
I really appreciate your idea and code — it works very well. Please don’t feel discouraged that I merged another PR for MPS support; in fact, I learned a lot from your implementation.
If I may suggest one thing, PR #67 is a good example of a style that makes it easier for us to review: every change is clear, effective, and easy to follow line by line.
I’m looking forward to your next contribution to VibeVoice. Thanks again for your efforts, and please stay tuned!
